Refactored library and radionet class

Emrvb opened the pull request:
- Added support for configurable api key
- Added better caching
- Added support for browsing additional categories
- Removed no longer supported API interactions
- Added a PlaybackProvider to mark tracks as live stream

Emrvb committed:
Removed no longer supported API interactions,

Emrvb committed:
Added a PlaybackProvider to mark tracks as live stream

Emrvb committed:
Refactored library and radionet class

- Changes in radio.net API (added support for configurable apikey)
- Added better caching
- Added support for browsing additional categories

Emrvb commented:
Maybe some people in https://github.com/plintx/mopidy-radionet/issues/16 can start toying around with this.

**Please note**

You need to add a valid api_key to your config file in the radionet section

kokosowy commented:
Wow, this is awesome work! I will try to test extension with your changes today.

Emrvb commented:
@kokosowy Thank you for the compliment.

There are a few more things I would like to fix, but I thought I should get the code out there quickly so people can comment.

tepiloxtl commented:
Got these two issues with your fork, using musicbox frontend: 

1. searching for any station yields "No tracks found...". There's `[RadioNetBackend-12] mopidy_radionet.radionet Radio.net: Found xx stations.` lines coming up in log each time I search, but each time searching with the same query makes number of stations increasing by... the amount of stations found I guess? (for "eska" it goes 17, 34, 51)
2. Ive added 'eska' as a favourite station slug (and it exists https://www.radio.net/s/eska), but in interface under favourite stations I get "KSKA-FM 91.1". I guess thats probably API issue though, and not plugin, but if it can be corrected somehow, would be nice

Emrvb commented:
@tepiloxtl 

1. I see I've indeed broken the search. Will fix that.
The increase in results is interesting. Good catch. I haven't really worked with python before and it seems the variables in the call stack are behaving a bit different from what I would have expected.

2. That's a pre-existing issue (https://github.com/plintx/mopidy-radionet/issues/15) but I'll see if I can fix that as well.

kokosowy commented:
Can anybody tell me step by step how I can fetch this pull request locally and then put it as python's module for Mopidy? Never done that before and github's guide is too much for my head I guess. Thanks.

Emrvb commented:
@kokosowy 

While **not recommended**, you could download the 5 files listed here: https://github.com/Emrvb/mopidy-radionet/tree/master/mopidy_radionet (when you click on them, look for the "raw" link) and use those to replace your local versions.

**Always backup local files before overwriting them**

tepiloxtl commented:
@kokosowy I just `git clone https://github.com/Emrvb/mopidy-radionet` and `sudo python3 -m pip install ./mopidy-radionet`

Edit: oh, not quite, I had to edit setup.py and edit `version=get_version('mopidy_radionet/__init__.py'),` to `version="0.2.2",`, because it didn't work from fork for some reason

kokosowy commented:
Thanks @Emrvb @tepiloxtl, I will back to the topic once I got some free time again. Good luck all.

anverx commented:
@Emrvb Could you explain the process of geting an API key?  I see nothing in mopidy logs that would look like a key.  There is nothing special in the request headers of the web interface.  I suppose for the web interface you don't need anything but to use and API one needs a key that is issued somehow?

Emrvb commented:
@anverx Easiest way to get a valid api key is to browse to radio.net and browse to a station. Click on the "Embed" widget. In the code that is shown, look for data-apikey. Put the value of that attribute in your mopidy config in the radionet section.

Emrvb committed:
Fixed regression in search

anverx commented:
zomg, it was so easy!  And at least browsing the lists is working, and i think playback as well.  Now i just need to sort analog audio out on my Allwinner A20 board.

Emrvb committed:
Fixed favorites inconsistent results

Should look up station by slug instead of search for station

Emrvb committed:
Populate station list from search results to prevent unnecessary api requests

Emrvb committed:
Don't list pages if there is just a single page

Emrvb commented:
@tepiloxtl

Both issues should be fixed, could you give it a try?

Edit:

For your second issue, watch your logs. If there are errors about stations not found, you have either the wrong slug, or the slug in quotes in your config file. In that case we fall back on search like it was done previously, with inconsistent results.

MiKiL75 commented:
@Emrvb Just wanted to drop by to say that I installed your 2f12cff and after some testing with different channels it works fine.
Going to do some more testing today.

tepiloxtl commented:
@Emrvb Search is fixed, but searching for `eska` still don't return the station Im looking for :P Would be nice if we could at least add stations to queue by slug using uri?
Favourites are working after I're removed quotes from config file.

On a side note, installing via pip from your repo trips because pip cant read 'version' string from setup.py. I've said that in my previous comment, though now Im adressing this directly

MiKiL75 commented:
@Emrvb @tepiloxtl 
I did run mopidy in more verbose mode to test why `eska` doesn't show up in the `favorites`.

At the moment something is messing parsing the favorites-line.
If I use stationslugs with apostrophes, those also appear in API-queries.

**favorite_stations = 'bbcradio2','eska'**
```
DEBUG    2021-09-09 15:37:36,199 [15220:HttpServer] mopidy.http.handlers
  Received WebSocket message from 192.168.1.100: '{"method":"core.library.browse","params":{"uri":"radionet:favorites"},"jsonrpc":"2.0","id":540}'
DEBUG    2021-09-09 15:37:36,207 [15220:RadioNetBackend-10] urllib3.connectionpool
  Starting new HTTPS connection (1): api.radio.net:443
DEBUG    2021-09-09 15:37:36,537 [15220:RadioNetBackend-10] urllib3.connectionpool
  https://api.radio.net:443 "GET /info/v2/search/station?station=%27bbcradio2%27&apikey=<VALID_API_KEY> HTTP/1.1" 400 None
ERROR    2021-09-09 15:37:36,541 [15220:RadioNetBackend-10] mopidy_radionet.radionet
  Radio.net: Error on get station by id 'bbcradio2'. Error: {"errorCode":101,"status":"error"}
DEBUG    2021-09-09 15:37:36,637 [15220:RadioNetBackend-10] urllib3.connectionpool
  https://api.radio.net:443 "GET /info/v2/search/stationsonly?query=%27bbcradio2%27&pageindex=1&apikey=<VALID_API_KEY> HTTP/1.1" 200 None
DEBUG    2021-09-09 15:37:36,639 [15220:RadioNetBackend-10] mopidy_radionet.radionet
  Radio.net: Done search
DEBUG    2021-09-09 15:37:36,715 [15220:RadioNetBackend-10] urllib3.connectionpool
  https://api.radio.net:443 "GET /info/v2/search/station?station=%27eska%27&apikey=<VALID_API_KEY> HTTP/1.1" 400 None
ERROR    2021-09-09 15:37:36,718 [15220:RadioNetBackend-10] mopidy_radionet.radionet
  Radio.net: Error on get station by id 'eska'. Error: {"errorCode":101,"status":"error"}
DEBUG    2021-09-09 15:37:36,821 [15220:RadioNetBackend-10] urllib3.connectionpool
  https://api.radio.net:443 "GET /info/v2/search/stationsonly?query=%27eska%27&pageindex=1&apikey=<VALID_API_KEY> HTTP/1.1" 200 None
DEBUG    2021-09-09 15:37:36,824 [15220:RadioNetBackend-10] mopidy_radionet.radionet
  Radio.net: Done search
INFO     2021-09-09 15:37:36,825 [15220:RadioNetBackend-10] mopidy_radionet.radionet
  Radio.net: Loaded 2 favorite stations.
DEBUG    2021-09-09 15:37:36,828 [15220:HttpServer] mopidy.http.handlers
  Sent WebSocket message to 192.168.1.100: '{"jsonrpc": "2.0", "id": 540, "result": [{"__model__": "Ref", "uri": "radionet:station:3244", "name": "BBC Radio 2", "type": "track"}, {"__model__": "Ref", "uri": "radionet:station:129459", "name": "KSKA-FM 91.1", "type": "track"}]}'
DEBUG    2021-09-09 15:37:36,838 [15220:HttpServer] mopidy.http.handlers
  Received WebSocket message from 192.168.1.100: '{"method":"core.library.lookup","params":{"uris":["radionet:station:3244","radionet:station:129459"]},"jsonrpc":"2.0","id":541}'
DEBUG    2021-09-09 15:37:36,844 [15220:HttpServer] mopidy.http.handlers
  Sent WebSocket message to 192.168.1.100: '{"jsonrpc": "2.0", "id": 541, "result": {"radionet:station:3244": [{"__model__": "Track", "uri": "radionet:track:3244", "name": "BBC Radio 2", "artists": [{"__model__": "Artist", "name": "BBC Radio 2"}], "album": {"__model__": "Album", "uri": "radionet:station:3244", "name": "With a repertoire covering more than 40 years, Radio 2 plays the widest selection of music on the radio. / Europe / United Kingdom - London", "artists": [{"__model__": "Artist", "name": "BBC Radio 2"}]}, "comment": "With a repertoire covering more than 40 years, Radio 2 plays the widest selection of music on the radio."}], "radionet:station:129459": [{"__model__": "Track", "uri": "radionet:track:129459", "name": "KSKA-FM 91.1", "artists": [{"__model__": "Artist", "name": "KSKA-FM 91.1"}], "album": {"__model__": "Album", "uri": "radionet:station:129459", "name": " / North America / USA - Anchorage AK", "artists": [{"__model__": "Artist", "name": "KSKA-FM 91.1"}]}, "comment": ""}]}}'
```

**favorite_stations = bbcradio2,eska** (This line also properly returns information of eska-channel instead of KSKA-FM)
```
DEBUG    2021-09-09 15:39:20,954 [15245:HttpServer] mopidy.http.handlers
  Received WebSocket message from 192.168.1.100: '{"method":"core.library.browse","params":{"uri":"radionet:favorites"},"jsonrpc":"2.0","id":563}'
DEBUG    2021-09-09 15:39:20,962 [15245:RadioNetBackend-10] urllib3.connectionpool
  Starting new HTTPS connection (1): api.radio.net:443
DEBUG    2021-09-09 15:39:21,278 [15245:RadioNetBackend-10] urllib3.connectionpool
  https://api.radio.net:443 "GET /info/v2/search/station?station=bbcradio2&apikey=<VALID_API_KEY> HTTP/1.1" 200 None
DEBUG    2021-09-09 15:39:21,283 [15245:RadioNetBackend-10] mopidy_radionet.radionet
  Radio.net: Done get top stations list
DEBUG    2021-09-09 15:39:21,342 [15245:RadioNetBackend-10] urllib3.connectionpool
  https://api.radio.net:443 "GET /info/v2/search/station?station=eska&apikey=<VALID_API_KEY> HTTP/1.1" 200 None
DEBUG    2021-09-09 15:39:21,345 [15245:RadioNetBackend-10] mopidy_radionet.radionet
  Radio.net: Done get top stations list
INFO     2021-09-09 15:39:21,346 [15245:RadioNetBackend-10] mopidy_radionet.radionet
  Radio.net: Loaded 2 favorite stations.
DEBUG    2021-09-09 15:39:21,348 [15245:HttpServer] mopidy.http.handlers
  Sent WebSocket message to 192.168.1.100: '{"jsonrpc": "2.0", "id": 563, "result": [{"__model__": "Ref", "uri": "radionet:station:3244", "name": "BBC Radio 2", "type": "track"}, {"__model__": "Ref", "uri": "radionet:station:4371", "name": "Eska", "type": "track"}]}'
DEBUG    2021-09-09 15:39:21,354 [15245:HttpServer] mopidy.http.handlers
  Received WebSocket message from 192.168.1.100: '{"method":"core.library.lookup","params":{"uris":["radionet:station:3244","radionet:station:4371"]},"jsonrpc":"2.0","id":564}'
DEBUG    2021-09-09 15:39:21,360 [15245:HttpServer] mopidy.http.handlers
  Sent WebSocket message to 192.168.1.100: '{"jsonrpc": "2.0", "id": 564, "result": {"radionet:station:3244": [{"__model__": "Track", "uri": "radionet:track:3244", "name": "BBC Radio 2", "artists": [{"__model__": "Artist", "name": "BBC Radio 2"}], "album": {"__model__": "Album", "uri": "radionet:station:3244", "name": "With a repertoire covering more than 40 years, Radio 2 plays the widest selection of music on the radio. / Europe / United Kingdom - London", "artists": [{"__model__": "Artist", "name": "BBC Radio 2"}]}, "genre": "Pop, Rock", "comment": "With a repertoire covering more than 40 years, Radio 2 plays the widest selection of music on the radio."}], "radionet:station:4371": [{"__model__": "Track", "uri": "radionet:track:4371", "name": "Eska", "artists": [{"__model__": "Artist", "name": "Eska"}], "album": {"__model__": "Album", "uri": "radionet:station:4371", "name": "The popular radio station from Polen with an urban fresh hit mix! / Europe / Poland - Warsaw", "artists": [{"__model__": "Artist", "name": "Eska"}]}, "genre": "Hits, Top 40 & Charts", "comment": "The popular radio station from Polen with an urban fresh hit mix!"}]}}'

```

**EDIT:**
Just to be sure I also tested with that example configuration line, slugs with apostrophes and linebreaks.

 **favorite_stations =
   'bbcradio1'
   'bbcradio2'
   'thetrip'
   'nectarine'**

```
DEBUG    2021-09-09 16:29:04,169 [15449:HttpServer] mopidy.http.handlers
  Received WebSocket message from 192.168.1.100: '{"method":"core.library.browse","params":{"uri":"radionet:favorites"},"jsonrpc":"2.0","id":686}'
DEBUG    2021-09-09 16:29:04,177 [15449:RadioNetBackend-10] urllib3.connectionpool
  Starting new HTTPS connection (1): api.radio.net:443
DEBUG    2021-09-09 16:29:04,528 [15449:RadioNetBackend-10] urllib3.connectionpool
  https://api.radio.net:443 "GET /info/v2/search/station?station=%27bbcradio1%27&apikey=<VALID_API_KEY> HTTP/1.1" 400 None
ERROR    2021-09-09 16:29:04,532 [15449:RadioNetBackend-10] mopidy_radionet.radionet
  Radio.net: Error on get station by id 'bbcradio1'. Error: {"errorCode":101,"status":"error"}
DEBUG    2021-09-09 16:29:04,633 [15449:RadioNetBackend-10] urllib3.connectionpool
  https://api.radio.net:443 "GET /info/v2/search/stationsonly?query=%27bbcradio1%27&pageindex=1&apikey=<VALID_API_KEY> HTTP/1.1" 200 None
DEBUG    2021-09-09 16:29:04,636 [15449:RadioNetBackend-10] mopidy_radionet.radionet
  Radio.net: Done search
DEBUG    2021-09-09 16:29:04,715 [15449:RadioNetBackend-10] urllib3.connectionpool
  https://api.radio.net:443 "GET /info/v2/search/station?station=%27bbcradio2%27&apikey=<VALID_API_KEY> HTTP/1.1" 400 34
ERROR    2021-09-09 16:29:04,717 [15449:RadioNetBackend-10] mopidy_radionet.radionet
  Radio.net: Error on get station by id 'bbcradio2'. Error: {"errorCode":101,"status":"error"}
DEBUG    2021-09-09 16:29:04,771 [15449:RadioNetBackend-10] urllib3.connectionpool
  https://api.radio.net:443 "GET /info/v2/search/stationsonly?query=%27bbcradio2%27&pageindex=1&apikey=<VALID_API_KEY> HTTP/1.1" 200 564
DEBUG    2021-09-09 16:29:04,773 [15449:RadioNetBackend-10] mopidy_radionet.radionet
  Radio.net: Done search
DEBUG    2021-09-09 16:29:04,852 [15449:RadioNetBackend-10] urllib3.connectionpool
  https://api.radio.net:443 "GET /info/v2/search/station?station=%27thetrip%27&apikey=<VALID_API_KEY> HTTP/1.1" 400 None
ERROR    2021-09-09 16:29:04,854 [15449:RadioNetBackend-10] mopidy_radionet.radionet
  Radio.net: Error on get station by id 'thetrip'. Error: {"errorCode":101,"status":"error"}
DEBUG    2021-09-09 16:29:04,956 [15449:RadioNetBackend-10] urllib3.connectionpool
  https://api.radio.net:443 "GET /info/v2/search/stationsonly?query=%27thetrip%27&pageindex=1&apikey=<VALID_API_KEY> HTTP/1.1" 200 None
DEBUG    2021-09-09 16:29:04,959 [15449:RadioNetBackend-10] mopidy_radionet.radionet
  Radio.net: Done search
DEBUG    2021-09-09 16:29:05,037 [15449:RadioNetBackend-10] urllib3.connectionpool
  https://api.radio.net:443 "GET /info/v2/search/station?station=%27nectarine%27&apikey=<VALID_API_KEY> HTTP/1.1" 400 None
ERROR    2021-09-09 16:29:05,040 [15449:RadioNetBackend-10] mopidy_radionet.radionet
  Radio.net: Error on get station by id 'nectarine'. Error: {"errorCode":101,"status":"error"}
DEBUG    2021-09-09 16:29:05,143 [15449:RadioNetBackend-10] urllib3.connectionpool
  https://api.radio.net:443 "GET /info/v2/search/stationsonly?query=%27nectarine%27&pageindex=1&apikey=<VALID_API_KEY> HTTP/1.1" 200 516
DEBUG    2021-09-09 16:29:05,145 [15449:RadioNetBackend-10] mopidy_radionet.radionet
  Radio.net: Done search
INFO     2021-09-09 16:29:05,146 [15449:RadioNetBackend-10] mopidy_radionet.radionet
  Radio.net: Loaded 4 favorite stations.

```

kokosowy commented:
I wonder whether my reported issue is similar to 'eska' problem - https://github.com/plintx/mopidy-radionet/issues/15#issue-843700848

Emrvb commented:
@tepiloxtl I only fixed the slug search for favorites and search in general. I think the search function is using a weird API endpoint, I'll change that and see what happens. Regarding pip, I know **absolutely nothing** about python and its ecosystem. I'll see if I can fix it, but suggestions are appreciated here.

@MiKiL75 & @kokosowy 
Yeah, I mentioned this here https://github.com/plintx/mopidy-radionet/pull/18#issuecomment-915121241 and here https://github.com/plintx/mopidy-radionet/pull/18#issuecomment-915641785

The feature resolving the favorites was using the wrong API endpoint and as far as I can tell those quotes should have never been there in the config file in the first place. It could also be that mopidy changed its config parsing along the way. 

Currently we try to resolve the slug (only works when without quotes in the config), if that fails we fall back to the original search implementation, which has inconsistent results.
I'm considering trimming the favorite slug values as I feel like this "change" will cause people to raise issues for years to come.

MiKiL75 commented:
@Emrvb 
> Yeah, I mentioned this here [#18 (comment)](https://github.com/plintx/mopidy-radionet/pull/18#issuecomment-915121241) and here [#18 (comment)](https://github.com/plintx/mopidy-radionet/pull/18#issuecomment-915641785)

Oh, didn't see those ones, my bad, sorry about that.

Emrvb committed:
Made favorites optional

Emrvb committed:
Fixed setup.py version metadata regex

Emrvb committed:
Trim single quotes from favorite slugs

Emrvb committed:
Removed old refresh logic

Emrvb committed:
Added image support

Emrvb commented:
@tepiloxtl

The problem with pip install should be fixed now. 

The Eska stations are a bit of a mystery to me. Indeed they don't show up when searching, but I also can't seem to find them when I browse the categories. I'm affraid this is a sign we should expect this api to die all together in the near future.

Emrvb commented:
@MiKiL75 

No worries mate, favorites are now optional and quotes will be trimmed.

tepiloxtl commented:
@Emrvb Proper regex for version is `"__([a-z]+)__ = \"([^\"]+)\""` (note the double quote inside square brackets too). The one currently in the file still doesnt work for me, it just outputs everything thats after version string too)

Emrvb committed:
Fixed setup.py version metadata regex (2/2)

Emrvb commented:
@tepiloxtl whoops, sorry about that.

plintx commented:
Hello, I don't have enough free time for testing. Anyone checking if it's okay? Can i merge this pull request @kokosowy?

kokosowy commented:
Hi all.. sorry it took a while. I installed fresh instance of mopidy, installed original extension, replaced 5 files from this PR and finally make configuration changes as in README example.

Generally it looks OK, but I'd recommend implementing getting API key automatically instead of putting it to config file manually, before release of this version. It seems it worked that way last time and now it's not implemented.

Only issue I see is if you put any country code you will get this in Mopidy log, but module works anyway:

> wrz 28 13:28:59 nana mopidy[59244]: ERROR    [MainThread] root Radio.net not supported language: pl # or net, de, at, fr, pt, es, dk, se, it

This is of course not really related to this PR, but it would be also good to make code prepared in case API returns nothing or gibberish, like below while I was trying to put infamous https://www.radio.pl/s/radio357 into favourites, as for now it just crashes entire module:

> wrz 28 14:02:47 nana mopidy[70021]: ERROR    [RadioNetBackend-3] mopidy_radionet.radionet Radio.net: Error on get station by id radio357. Error: {"errorCode":101,"status":>
> wrz 28 14:02:47 nana mopidy[70021]: ERROR    [Core-7] mopidy.core.library RadioNetBackend backend caused an exception.
> wrz 28 14:02:47 nana mopidy[70021]: Traceback (most recent call last):
> wrz 28 14:02:47 nana mopidy[70021]:   File "/usr/lib/python3/dist-packages/mopidy/core/library.py", line 17, in _backend_error_handling
> wrz 28 14:02:47 nana mopidy[70021]:     yield
> wrz 28 14:02:47 nana mopidy[70021]:   File "/usr/lib/python3/dist-packages/mopidy/core/library.py", line 114, in _browse
> wrz 28 14:02:47 nana mopidy[70021]:     result = backend.library.browse(uri).get()
> wrz 28 14:02:47 nana mopidy[70021]:   File "/usr/lib/python3/dist-packages/pykka/_threading.py", line 45, in get
> wrz 28 14:02:47 nana mopidy[70021]:     _compat.reraise(*self._data['exc_info'])
> wrz 28 14:02:47 nana mopidy[70021]:   File "/usr/lib/python3/dist-packages/pykka/_compat/__init__.py", line 29, in reraise
> wrz 28 14:02:47 nana mopidy[70021]:     raise value
> wrz 28 14:02:47 nana mopidy[70021]:   File "/usr/lib/python3/dist-packages/pykka/_actor.py", line 193, in _actor_loop
> wrz 28 14:02:47 nana mopidy[70021]:     response = self._handle_receive(envelope.message)
> wrz 28 14:02:47 nana mopidy[70021]:   File "/usr/lib/python3/dist-packages/pykka/_actor.py", line 299, in _handle_receive
> wrz 28 14:02:47 nana mopidy[70021]:     return callee(*message.args, **message.kwargs)
> wrz 28 14:02:47 nana mopidy[70021]:   File "/usr/local/lib/python3.8/dist-packages/mopidy_radionet/library.py", line 68, in browse
> wrz 28 14:02:47 nana mopidy[70021]:     return self._browse_category(category, page)
> wrz 28 14:02:47 nana mopidy[70021]:   File "/usr/local/lib/python3.8/dist-packages/mopidy_radionet/library.py", line 109, in _browse_category
> wrz 28 14:02:47 nana mopidy[70021]:     items = self._get_favorites()
> wrz 28 14:02:47 nana mopidy[70021]:   File "/usr/local/lib/python3.8/dist-packages/mopidy_radionet/library.py", line 219, in _get_favorites
> wrz 28 14:02:47 nana mopidy[70021]:     return self.backend.radionet.get_favorites()
> wrz 28 14:02:47 nana mopidy[70021]:   File "/usr/local/lib/python3.8/dist-packages/mopidy_radionet/radionet.py", line 375, in get_favorites
> wrz 28 14:02:47 nana mopidy[70021]:     json["categories"][0]["matches"][0]
> wrz 28 14:02:47 nana mopidy[70021]: IndexError: list index out of range

Thanks for all your dev work, at least I can participate testing it and submitting issues. :-)

MiKiL75 commented:
> Only issue I see is if you put any country code you will get this in Mopidy log, but module works anyway:
> 
> > wrz 28 13:28:59 nana mopidy[59244]: ERROR    [MainThread] root Radio.net not supported language: pl # or net, de, at, fr, pt, es, dk, se, it
> 

Did you use language line directly without trimming commented section off like:
`language = pl # or net, de, at, fr, pt, es, dk, se, it`
At the moment you have to manually remove rest off leaving only one language on the line.

Emrvb commented:
I think it should default to .net if it doesn't recognise the suffix. Edit: it does that already, if the problem is what MiKiL describes, I guess I could fix that instead.

I'll have a look at Radio 357, but keep in mind, this old api seems to be missing stations.

kokosowy commented:
> At the moment you have to manually remove rest off leaving only one language on the line.

@MiKiL75 yup, you're absolutely right. Didn't know Mopidy can't handle these kind of comments in config file. Thanks. So it's good.

> 'll have a look at Radio 357, but keep in mind, this old api seems to be missing stations.

@Emrvb it is most probably API missing it, but still it'd be great if code would have some kind of damage control for exceptions like this. :-)

And this API key thing, I still think it should be fetched automatically and it's most important for now.

Emrvb committed:
Fixed error caused by unrecognized favorite not yielding result in search

Emrvb committed:
Fixed some oddities with language setting

- Strip whitespaces from setting
- Accept "en" as alternative for "net"
- Log warning instead of error

MiKiL75 commented:
Open issue in mopidy:
Inline config comments broken in Mopidy 3 https://github.com/mopidy/mopidy/issues/1868

Emrvb committed:
Support for track uri with station slug instead of id

Emrvb committed:
Code cleanup

Emrvb committed:
Removed lingering logging statement and fixed caching issue causing tests to fail

Emrvb commented:
@kokosowy 

The crash with the unrecognized station slug should be fixed now

Regarding the API key: I don't really feel comfortable fetching the API key from the javascripts as it would be pretty suspectible to breaking.

kokosowy commented:
@Emrvb wow, thanks for commits, tested, seems great! Of course I completely respect your opinion regarding obtaining the API key, but what would you say if the script could obtain the key even from js and it would be also still possible to use optional parameter api_key which would mean using api_key value and not obtaining the API key from the js? I understand your reasons, however I'm representing here typical users and the way of obtaining API key automatically, even from a shady js file and having a possibility of using api_key parameter in case of issues, is the way I'd definitely prefer. But of course I'm still happy what we already have here thanks to you, thanks a dozen!

Emrvb commented:
@kokosowy, you're very welcome, no problem.

Regarding the API key, there is also the question whether we even should automatically download the api key. It could be considered circumventing a digital access restriction. It also could trigger radio.de to close this loophole. Another consideration is that it's better to have users know they need an API key and how to get it, in case it fails. In the end I guess it's up to @plintx if he wants something like that in his addon.

Btw, you don't need to go through the js file. The easiest way to get a valid api key is to browse to radio.net and browse to a station. Click on the "Embed" widget. In the code that is shown, look for data-apikey. Put the value of that attribute in your mopidy config in the radionet section.

kokosowy commented:
How are you doing guys? Just asking. :-)

Emrvb commented:
@kokosowy I'm fine, hope you are too. I don't think plintx wants to put much thought into to this. So you should probably let him know if you think this should be merged or not.

kokosowy commented:
Hi all!

Thanks @Emrvb. I'm still new in this process.

@plintx this PR definitely fix the issue, so I'd say we can merge. You just need to be aware API key needs to be fetched manually now on and put into configuration file. README.rst will need an update, I can take care of it once merged.

kokosowy commented:
Hi @plintx, how are you doing? Got some time? :-)

ghtmtt commented:
Hi all! I'm trying this PR but it seems something is not working here.

I cloned the repository, checked out this PR (to be sure also did a `cat` on /usr/local/lib/python3/dist-packages/mopidy_radionet/__init__.py`) but the content in iris is blank, for example the favorites: 

![image](https://user-images.githubusercontent.com/2884884/148410340-833c8117-b4da-4de9-b8b5-c508ea10856c.png)


There are all the command that I ran:

```
git clone git@github.com:plintx/mopidy-radionet.git
git fetch origin pull/18/head:radio
git checkout radio

sudo python3 -m pip install ./mopidy-radionet/
```

Finally this is what I add to the `/etc/mopidy/mopidy.config` file (I'm running mopidy as a service):

```
[radionet]
enabled = true
language = it
min_bitrate = 96
favorite_stations =
  rairadiodue
```
Am I missing something weird?

BTW: I'm **really** willingly to help with this PR!

kokosowy commented:
Hi! I think you're missing an api key parameter. Read whole thread for details. In case you're still having issues logs of mopidy service also help.

ghtmtt commented:
@kokosowy ahhhrrrrrr shame on me!! gotcha now and working fine! I read the whole tread but I should have missed that comment :) Thanks!!

kokosowy commented:
I'm glad it works for you. We should really apply this PR and update documentation. :-)

ghtmtt commented:
if I might give a suggestion, rename the parameter `api_key` to `apikey`. Just because in the embedded widget the name is `apikey`. 

> I'm glad it works for you. We should really apply this PR and update documentation. :-)

**Definitely!!**

ChrisZ32 commented:
This branch worked fine for me too, thanks! To collate the steps that helped me (especially if you aren't familiar with git):

1) Installation - Clone this branch from git to a folder (you will need to setup git on your machine/rpi first, see https://docs.github.com/en/authentication/troubleshooting-ssh/error-permission-denied-publickey)
- run "git clone git@github.com:plintx/mopidy-radionet.git" from the folder you want to clone into
- cd into the newly created "mopidy-radionet" folder
- run "git fetch origin pull/18/head:radio"
- run "git checkout radio"
- cd one level up with "cd ../", back to the original folder
- run "sudo python3 -m pip install ./mopidy-radionet/" to install

2) Config
- Setup normal radio.net config (already detailed in the main readme [here)](https://github.com/plintx/mopidy-radionet#configuration)
- Get an api key by browsing radio.net website to any stream, click the "Embed" widget to get a popup, in the code displayed look for a section that says "data-apikey" and copy the long text string after it
- Add this new key to the radio.net configuration section  as "api_key KEYYOUCOPIEDHERE"
- Restart mopidy

That should be it!

natumbri commented:
Thanks for all the work on this - it works great for me, except that some stations don't seem to return a streamUrl, even though they stream just fine from the `radio.net` website.  Radio station `3rrr` is an example.  The stream is at `https://realtime.rrr.org.au/p13`.  The `response.json()` in `_get_station_by_id` for `3rrr` is this:

```
{
    "continent": "Australia",
    "country": "Australia",
    "logo300x300": "https://static.radio.net/images/broadcasts/a4/2b/7870/c300.png",
    "city": "Melbourne",
    "stationType": "radio_station",
    "description": "The independent radio station from Melbourne offering you a wide selection of music-pleasure. The playlist includes Indie Rock, Blues and Folk and inbetween Interviews with artists to keep the program entertaining.",
    "language": [
        "English"
    ],
    "logo100x100": "https://static.radio.net/images/broadcasts/a4/2b/7870/c100.png",
    "streamUrls": [],
    "playable": "PLAYABLE",
    "genres": [
        "Pop",
        "Indie",
        "Hits"
    ],
    "logo175x175": "https://static.radio.net/images/broadcasts/a4/2b/7870/c175.png",
    "adParams": {
        "st_city": [
            "Melbourne"
        ],
        "languages": [
            "English"
        ],
        "genres": [
            "Hits",
            "Indie",
            "Pop"
        ],
        "topics": [],
        "st_cont": [
            "Australia"
        ],
        "station": [
            "3rrr"
        ],
        "advertising": [
            "true"
        ],
        "family": [],
        "st_region": [
            "Victoria"
        ],
        "type": [
            "radio_station"
        ],
        "st_country": [
            "Australia"
        ]
    },
    "alias": "",
    "rank": 4216,
    "id": 7870,
    "types": [
        "FM Radio"
    ],
    "website": "https://www.rrr.org.au/",
    "topics": [],
    "shortDescription": "The playlist includes Indie Rock, Blues and Folk and inbetween Interviews with artists to keep the program entertaining.",
    "logo44x44": "https://static.radio.net/images/broadcasts/a4/2b/7870/c44.png",
    "numberEpisodes": 0,
    "podcastUrls": [],
    "hideReferer": false,
    "name": "3RRR Triple R 102.7 FM",
    "subdomain": "3rrr",
    "lastModified": "2022-06-28T05:03:56.000Z",
    "family": [],
    "region": "Victoria",
    "frequencies": []
}
```

Emrvb commented:
@natumbri 

They're sunsetting the API we're using. Some stations dont show up on search, others like this apparently have their stream urls not set.

MiKiL75 commented:
I was trying to look their new api-code how to list station information, search stations etc.

Here something what I found out by using developer tools.
All the relevant info for building urls is found from at the moment `_next/static/chunks/pages/_app-d75cc82e439caa15.js`
new api prefix url is: https://prod.radio-api.net/

Api query-paths and parameters directly from code:
```
		}, {
			key: "getBroadcast",
			value: function(e, t) {
				return this.request("/stations/legacy/details?broadcastId=".concat(e), t)
			}
		}, {
			key: "getBroadcastMulti",
			value: function(e, t) {
				return this.request("/stations/details?stationIds=".concat(e), t)
			}
		}, {
			key: "getPodcastEpisodes",
			value: function(e, t) {
				var n = arguments.length > 2 && void 0 !== arguments[2] ? arguments[2] : 5,
					r = arguments.length > 3 && void 0 !== arguments[3] ? arguments[3] : 0;
				return this.request("/podcasts/episodes/by-podcast-ids?podcastIds=".concat(e, "&count=").concat(n, "&offset=").concat(r), t)
			}
		}, {
			key: "getNowPlaying",
			value: function(e) {
				return this.request("/stations/".concat(e, "/songs"))
			}
		}, {
			key: "getSimilarStations",
			value: function(e, t) {
				var n = arguments.length > 2 && void 0 !== arguments[2] ? arguments[2] : 12;
				return this.request("/stations/legacy/".concat(e, "/similar?count=").concat(n), t)
			}
		}, {
			key: "getPodcastsByFamily",
			value: function(e, t) {
				var n = arguments.length > 2 && void 0 !== arguments[2] ? arguments[2] : 15,
					r = arguments.length > 3 && void 0 !== arguments[3] ? arguments[3] : 0;
				return this.request("/stations/legacy/".concat(e, "/family/podcasts?count=").concat(n, "&offset=").concat(r), t)
			}
		}, {
			key: "getStationsByFamily",
			value: function(e, t) {
				var n = arguments.length > 2 && void 0 !== arguments[2] ? arguments[2] : 15,
					r = arguments.length > 3 && void 0 !== arguments[3] ? arguments[3] : 0;
				return this.request("/stations/legacy/".concat(e, "/family/stations?count=").concat(n, "&offset=").concat(r), t)
			}
		}, {
			key: "getStationsBySystemname",
			value: function(e, t) {
				var n = arguments.length > 2 && void 0 !== arguments[2] ? arguments[2] : 10,
					r = arguments.length > 3 && void 0 !== arguments[3] ? arguments[3] : 0;
				return this.request("/stations/list-by-system-name?systemName=".concat(e, "&count=").concat(n, "&offset=").concat(r), t)
			}
		}, {
			key: "getSystemTags",
			value: function(e) {
				return this.request("/stations/tags", e)
			}
		}, {
			key: "getStationsByTag",
			value: function(e, t, n) {
				var r = arguments.length > 3 && void 0 !== arguments[3] ? arguments[3] : 50,
					o = arguments.length > 4 && void 0 !== arguments[4] ? arguments[4] : 0;
				return this.request("/stations/legacy/by-tag?tagType=".concat(t, "&slug=").concat(e, "&count=").concat(r, "&offset=").concat(o), n)
			}
		}, {
			key: "getStationsByLetter",
			value: function(e, t) {
				var n = arguments.length > 2 && void 0 !== arguments[2] ? arguments[2] : 200,
					r = arguments.length > 3 && void 0 !== arguments[3] ? arguments[3] : 0;
				return this.request("/stations/".concat(e, "?count=").concat(n, "&offset=").concat(r), t)
			}
		}, {
			key: "getShortlist",
			value: function(e, t) {
				var n = arguments.length > 2 && void 0 !== arguments[2] ? arguments[2] : 20;
				return this.request("/stations/shortlist?tagType=".concat(e, "&count=").concat(n), t)
			}
		}, {
			key: "getStationsBySearch",
			value: function(e, t, n, r) {
				return this.request("/stations/legacy/search?query=".concat(e, "&count=").concat(n, "&offset=").concat(r), t)
			}
		}, {
			key: "getRelevantStations",
			value: function(e) {
				var t = arguments.length > 1 && void 0 !== arguments[1] ? arguments[1] : 1e3,
					n = arguments.length > 2 && void 0 !== arguments[2] ? arguments[2] : 0;
				return this.request("/stations/sitemap/ids?count=".concat(t, "&offset=").concat(n, "&seoRelevantOnly=true"), e)
			}
		}, {
			key: "getFrequenciesByCity",
			value: function() {
				var e = (0, c.Z)(u().mark((function e(t, n) {
					var r;
					return u().wrap((function(e) {
						for (;;) switch (e.prev = e.next) {
							case 0:
								return e.prev = 0, e.next = 3, this.request("/stations/cities/".concat(t, "/frequencies"), n);
							case 3:
								r = e.sent, e.next = 9;
								break;
							case 6:
								e.prev = 6, e.t0 = e.catch(0), r = [];
							case 9:
								return e.abrupt("return", r);
							case 10:
							case "end":
								return e.stop()
						}
					}), e, this, [
						[0, 6]
					])
				})));
				return function(t, n) {
					return e.apply(this, arguments)
				}
			}()
		}], [{
```
These two gives station information, stream url etc. This takes the station's slug as input.
```
/stations/legacy/details?broadcastId=3rrr
/stations/details?stationIds=ylepuhe
```

List of all the usable language-, topic-, country-, genre-, city-tags
`/stations/tags`

Searching stations by using those tags for example:
```
/stations/legacy/by-tag?tagType=language&slug=French&count=50
/stations/legacy/by-tag?tagType=topic&slug=Arts&count=50
/stations/legacy/by-tag?tagType=country&slug=Japan&count=50&offset=1
/stations/legacy/by-tag?tagType=genre&slug=Rock&count=50&offset=1
/stations/legacy/by-tag?tagType=city&slug=Paris&count=50
```

Searching stations by letter
`/stations/c?count=100&offset=1`

Normal station search by word (path without legacy-part wasn't in the code but it works too)
```
/stations/legacy/search?query=
/stations/search?query=
```

Emrvb commented:
@MiKiL75 

Well that's funny. When I worked on this I checked the website to see if I could use the new API, but I never saw it use any API at all. Could you perhaps provide a copy of a working request? I would like to see the headers they use. Raw from the inspector is fine, otherwise a "Copy as cURL". I'm pretty busy at the moment but I might find some time to look into this.

MiKiL75 commented:
@Emrvb 
No worries if you're busy at the moment. Surprisingly old api is still working after a year.

Here's few taken with Copy as cURL (bash). I disabled cache while I took these so that seems to be a reason they have have
`cache-control: no-cache` and `pragma: no-cache`

_Edited a bit, replaced Finnish radio station that had old non-working url to stream._
_Note: tagged searching doesn't have sorting option (By rank or Alphabetical) what old version has._

1. get the tag-list.
2. normal search for word 3rrr.
3. legacy details search for station with slug-id 3rrr.
4. new details search for station with slug-id aalto.
5. search for stations tagged language:Japanese with count 50 per page.
6. search for stations tagged topic:Arts with count 50 per page.
7. search for stations tagged genre:2000s with count 50 per page.
8. search station names starting with letter 'e' with count 50 per page offsetted by 3.

```
curl 'https://prod.radio-api.net/stations/tags' \
  -H 'authority: prod.radio-api.net' \
  -H 'accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9' \
  -H 'cache-control: no-cache' \
  -H 'pragma: no-cache' \
  -H 'sec-ch-ua: "Chromium";v="102", " Not A;Brand";v="99"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Windows"' \
  -H 'sec-fetch-dest: document' \
  -H 'sec-fetch-mode: navigate' \
  -H 'sec-fetch-site: none' \
  -H 'sec-fetch-user: ?1' \
  -H 'upgrade-insecure-requests: 1' \
  -H 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.149 Safari/537.36' \
  --compressed

curl 'https://prod.radio-api.net/stations/search?query=3rrr' \
  -H 'authority: prod.radio-api.net' \
  -H 'accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9' \
  -H 'cache-control: no-cache' \
  -H 'pragma: no-cache' \
  -H 'sec-ch-ua: "Chromium";v="102", " Not A;Brand";v="99"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Windows"' \
  -H 'sec-fetch-dest: document' \
  -H 'sec-fetch-mode: navigate' \
  -H 'sec-fetch-site: none' \
  -H 'sec-fetch-user: ?1' \
  -H 'upgrade-insecure-requests: 1' \
  -H 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.149 Safari/537.36' \
  --compressed

curl 'https://prod.radio-api.net/stations/legacy/details?broadcastId=3rrr' \
  -H 'authority: prod.radio-api.net' \
  -H 'accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9' \
  -H 'cache-control: no-cache' \
  -H 'pragma: no-cache' \
  -H 'sec-ch-ua: "Chromium";v="102", " Not A;Brand";v="99"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Windows"' \
  -H 'sec-fetch-dest: document' \
  -H 'sec-fetch-mode: navigate' \
  -H 'sec-fetch-site: none' \
  -H 'sec-fetch-user: ?1' \
  -H 'upgrade-insecure-requests: 1' \
  -H 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.149 Safari/537.36' \
  --compressed

curl 'https://prod.radio-api.net/stations/details?stationIds=aalto' \
  -H 'authority: prod.radio-api.net' \
  -H 'accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9' \
  -H 'cache-control: no-cache' \
  -H 'pragma: no-cache' \
  -H 'sec-ch-ua: "Chromium";v="102", " Not A;Brand";v="99"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Windows"' \
  -H 'sec-fetch-dest: document' \
  -H 'sec-fetch-mode: navigate' \
  -H 'sec-fetch-site: none' \
  -H 'sec-fetch-user: ?1' \
  -H 'upgrade-insecure-requests: 1' \
  -H 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.149 Safari/537.36' \
  --compressed

curl 'https://prod.radio-api.net/stations/legacy/by-tag?tagType=language&slug=Japanese&count=50' \
  -H 'authority: prod.radio-api.net' \
  -H 'accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9' \
  -H 'cache-control: no-cache' \
  -H 'pragma: no-cache' \
  -H 'sec-ch-ua: "Chromium";v="102", " Not A;Brand";v="99"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Windows"' \
  -H 'sec-fetch-dest: document' \
  -H 'sec-fetch-mode: navigate' \
  -H 'sec-fetch-site: none' \
  -H 'sec-fetch-user: ?1' \
  -H 'upgrade-insecure-requests: 1' \
  -H 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.149 Safari/537.36' \
  --compressed

curl 'https://prod.radio-api.net/stations/legacy/by-tag?tagType=topic&slug=Arts&count=50' \
  -H 'authority: prod.radio-api.net' \
  -H 'accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9' \
  -H 'cache-control: no-cache' \
  -H 'pragma: no-cache' \
  -H 'sec-ch-ua: "Chromium";v="102", " Not A;Brand";v="99"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Windows"' \
  -H 'sec-fetch-dest: document' \
  -H 'sec-fetch-mode: navigate' \
  -H 'sec-fetch-site: none' \
  -H 'sec-fetch-user: ?1' \
  -H 'upgrade-insecure-requests: 1' \
  -H 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.149 Safari/537.36' \
  --compressed

curl 'https://prod.radio-api.net/stations/legacy/by-tag?tagType=genre&slug=2000s&count=50' \
  -H 'authority: prod.radio-api.net' \
  -H 'accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9' \
  -H 'cache-control: no-cache' \
  -H 'pragma: no-cache' \
  -H 'sec-ch-ua: "Chromium";v="102", " Not A;Brand";v="99"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Windows"' \
  -H 'sec-fetch-dest: document' \
  -H 'sec-fetch-mode: navigate' \
  -H 'sec-fetch-site: none' \
  -H 'sec-fetch-user: ?1' \
  -H 'upgrade-insecure-requests: 1' \
  -H 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.149 Safari/537.36' \
  --compressed

curl 'https://prod.radio-api.net/stations/e?count=50&offset=3' \
  -H 'authority: prod.radio-api.net' \
  -H 'accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9' \
  -H 'cache-control: no-cache' \
  -H 'pragma: no-cache' \
  -H 'sec-ch-ua: "Chromium";v="102", " Not A;Brand";v="99"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Windows"' \
  -H 'sec-fetch-dest: document' \
  -H 'sec-fetch-mode: navigate' \
  -H 'sec-fetch-site: none' \
  -H 'sec-fetch-user: ?1' \
  -H 'upgrade-insecure-requests: 1' \
  -H 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.149 Safari/537.36' \
  --compressed
```

Some things I haven't got working, getStationsBySystemname.
`/stations/list-by-system-name?systemName=Afrikaans`
Afrikaans should be valid systemname according to that list but it returns error `unsupported system name`

Emrvb commented:
@MiKiL75 

Thanks for the examples. I'm a bit confused all these request are valid without authentication.

Regarding the old API, there is some history there. I think they feel obligated to keep it online until they no longer technically can without a substantial investment.

GodotIsWaitingToo commented:
I've put the below in as the config text, but it's giving me a message when I interrogate the config (sudo mopidyctl config) that says 'Extension disabled due to config errors'. I installed RadioNet using sudo pip3 install Mopidy-RadioNet. Any help much appreciated!

[radionet]
enabled = true
language = net
api_key = df04ff67dd3339a6fc19c9b8be164d5b5245ae93
min_bitrate = 96
favorite_stations =
  'bbcradio4'

r3k2 commented:
Need help on this, how does one register from an api key? I was not able to see it
